src/shell/shell_cmds.cpp

```diff
@@ -2242,3 +2243,298 @@ void DOS_Shell::CMD_VOL(char* args)
+			}
+		} while (DOS_FindNext());
+	}
+
+	// Done with DTA. Restore it.
+	dos.dta(save_dta);
+
+	uint16_t destination_attr = 0;
+	bool destination_exists = DOS_GetFileAttr(canonical_destination.c_str(),
+	                                          &destination_attr);
+	bool dest_is_dir       = false;
+
+	if (destination_exists) {
+		if (destination_attr & DOS_ATTR_DIRECTORY) {
+			dest_is_dir = true;
+		} else if (final_sources.size() > 1) {
+			WriteOut(MSG_Get("SHELL_CMD_MOVE_MULTIPLE_TO_SINGLE"));
+			return;
+		}
+	} else if (final_sources.size() > 1) {
+		dest_is_dir = true;
+		if (!DOS_MakeDir(canonical_destination.c_str())) {
+			WriteOut(MSG_Get("SHELL_CMD_MKDIR_ERROR"),
+			         canonical_destination.c_str());
+			return;
+		}
+	}
+
+	if (dest_is_dir && canonical_destination.back() != DosSeparator) {
+		canonical_destination.push_back(DosSeparator);
+	}
+
+	for (const std::string& source : final_sources) {
+		std::string final_destination = canonical_destination;
+		if (dest_is_dir) {
+			final_destination += source.substr(
+			        source.rfind(DosSeparator) + 1);
+		}
+		if (source == final_destination) {
+			WriteOut(MSG_Get("SHELL_FILE_EXISTS"),
+			         final_destination.c_str());
+			continue;
+		}
+		// Overwrite existing file
+		if (destination_exists && !dest_is_dir) {
+			// Condition should be checked for previously
+			assert(final_sources.size() == 1);
+			if (!DOS_UnlinkFile(final_destination.c_str())) {
+				WriteOut(MSG_Get("SHELL_CMD_DEL_ERROR"),
+				         final_destination.c_str());
+				return;
+			}
+		}
+		// If same drive, do a rename. Otherwise it needs a copy + delete.
+		if (source[0] == final_destination[0]) {
+			if (DOS_Rename(source.c_str(), final_destination.c_str())) {
+				WriteOut("%s => %s\n",
+				         source.c_str(),
+				         final_destination.c_str());
+			} else {
+				WriteOut(MSG_Get("SHELL_FILE_CREATE_ERROR"),
+				         final_destination.c_str());
+			}
+		} else {
+			uint16_t source_attr = 0;
+			if (!DOS_GetFileAttr(source.c_str(), &source_attr)) {
+				WriteOut("SHELL_FILE_NOT_FOUND", source.c_str());
+				continue;
+			}
+			uint16_t source_handle = 0;
+			if (!DOS_OpenFile(source.c_str(), OPEN_READ, &source_handle)) {
+				WriteOut("SHELL_FILE_OPEN_ERROR", source.c_str());
+				continue;
+			}
+			uint16_t dest_handle = 0;
+			if (!DOS_CreateFile(final_destination.c_str(),
+			                    source_attr,
+			                    &dest_handle)) {
+				WriteOut(MSG_Get("SHELL_FILE_CREATE_ERROR"),
+				         final_destination.c_str());
+				DOS_CloseFile(source_handle);
+				continue;
+			}
+			constexpr uint16_t buffer_capacity = 4096;
+			uint8_t buffer[buffer_capacity];
+			uint16_t bytes_requested = buffer_capacity;
+			bool success             = true;
+			do {
+				if (!DOS_ReadFile(source_handle, buffer, &bytes_requested)) {
+					WriteOut(MSG_Get("SHELL_READ_ERROR"),
+					         source.c_str());
+					success = false;
+					break;
```

Comment:
As there's nothing more to do in this function after this, I'd prefer a `return` here.